doc/design/008-state-proof-pluggable-parsing/README.md

```diff
@@ -0,0 +1,127 @@
+
+## Goals
+Libindy allows to create and send custom transaction via plugged interface.
+It is nice to have a way to support BLS MS and SP verification for these plugged transactions.
+
+Implementation of math for SP verification is a bit complicate for plugin logic.
+Therefore libindy should perform all math calculation inside.
+A plugin should provide handler to parse custom reply to fixed data structure.
+
+## API
+The signature of the handler is described below together with custom `free` call to deallocate result data.
+
+```rust
+extern fn custom_transaction_parser(reply_from_node: *const c_char, parsed_sp: *mut *const c_char) -> ErrCode;
+extern fn custom_free(data: *mut c_char) -> ErrCode;
+``` 
+
+Libindy API will contain call to register handler for specific transaction type:
+```rust
+extern fn indy_register_transaction_parser_for_sp(command_handle: i32,
+                                                  txn_type: *const c_char,
+                                                  parser: custom_transaction_parser,
+                                                  free: custom_free,
+                                                  cb: extern fn(command_handle_: i32, err: ErrCode)) -> ErrCode;
+```
+
+### Parsed Data structure
+
+A plugin should parse `reply_from_node` and return back to libindy parsed data as JSON string.
+Actually this data is array of entities each of them is describe SP Trie and set of key-value pairs to verify against this trie.
+It can be represented as `Vec<ParsedSP>` serialized to JSON.
+
+
+```rust
+/** 
+ Single item to verification:
+ - SP Trie with RootHash
+ - BLS MS 
+ - set of key-value to verify
+*/ 
+struct ParsedSP {
+    proof_nodes: String, /// encoded SP Trie transferred from Node to Client
+    root_hash: String, /// RootHash of the Trie, start point for verification. Should be same with appropriate filed in BLS MS data 
+    kvs_to_verify: KeyValuesInSP, /// entities to verification against current SP Trie
+    multi_signature: JsonObject, /// BLS MS data for verification
+}
+
+/**
+ Variants of representation for items to verify against SP Trie
+ Right now 2 options are specified:
+ - simple array of key-value pair
+ - whole subtrie
+*/
+enum KeyValuesInSP {
+    Simple(KeyValueSimpleData),
+    SubTrie(KeyValuesSubTrieData),
+}
+
+/**
+ Simple variant of `KeyValuesInSP`.
+ 
+ All required data already present in parent SP Trie (built from `proof_nodes`).
+ `kvs` can be verified directly in parent trie 
+*/
+struct KeyValueSimpleData {
+    kvs: Vec<(String /* key */, String /* val */)>
+}
+
+/**
+ Subtrie variant of `KeyValuesInSP`.
+ 
+ In this case Client (libindy) should construct subtrie and append it into trie based on `proof_nodes`.
+ After this preparation each kv pair can be checked.
+*/
+struct KeyValuesSubTrieData {
+    sub_trie_prefix: Option<String>, /// common prefix of each pair in `kvs`. Should be used to correct merging initial trie and subtrie
```

Comment:
Why does it care that there is a prefix? The node returns the full keys as of now. It does not mention the prefix? Are you thinking of the optimisation that we planned for later?

Reply:
yes, SubTrie option is just reserve for future optimization. In the fist iteration all components (Node, libindy, plugins) will not support this variant.